File: r/R/cde-loss.R
#' Calculate conditional density estimation loss.
#'
#' Estimates the CDE loss
#' \deqn{\int \int (\hat{f}(z \mid x) - f(z \mid x))^{2} dz dP(x)}
#' which can be written as
#' \deqn{E[\int \hat{f}(z \mid X) dz] - 2 E[\hat{f}(Z \mid X)] + C_{f}}
#' where \eqn{C_{f}} is an (unknown) constant. These expectations are
#' estimated using empirical expectations.
#'
#' @param cdes a matrix of conditional density estimates with each row
#'   corresponding to an observation, and each column corresponding to
#'   a grid point
#' @param z_grid a matrix of the grid points at which the conditional
#'   densities are estimated
#' @param z_test a matrix of the true responses
#' @examples
#' z_grid <- seq(0, 1, length.out = 99)
#' z_test <- runif(101)
#' cde <- matrix(dbeta(z_grid, 2, 2), nrow = 101, ncol = 99, byrow=TRUE)
#' cde_loss(cde, z_grid, z_test)
#' @export
cde_loss <- function(cdes, z_grid, z_test) {
  if (!is.matrix(z_grid)) {
    z_grid <- as.matrix(z_grid)
  }

  if (!is.matrix(z_test)) {
    z_test <- as.matrix(z_test)
  }

  stopifnot(nrow(cdes) == nrow(z_test))
  stopifnot(ncol(cdes) == nrow(z_grid))
  stopifnot(ncol(z_grid) == ncol(z_test))

  z_min <- apply(z_grid, 2, min)
  z_max <- apply(z_grid, 2, max)
  z_delta <- prod(z_max - z_min) / nrow(z_grid)

  integrals <- z_delta * rowSums(cdes ^ 2)

  nn_ids <- cbind(seq_len(nrow(z_test)), FNN::knnx.index(z_grid, z_test, k = 1))
  likeli <- cdes[nn_ids]

  losses <- integrals - 2 * likeli

  return(list(loss = mean(losses),
              se = stats::sd(losses) / sqrt(nrow(cdes))))
}

File: python/cde_loss.py
import numpy as np
import scipy.stats as stats

def cde_loss(cde_estimates, z_grid, z_test):
    """Calculates conditional density estimation loss on holdout data

    @param cde_estimates: a numpy array where each row is a density
    estimate on z_grid
    @param z_grid: a numpy array of the grid points at which cde_estimates is evaluated
    @param z_test: a numpy array of the true z values corresponding to the rows of cde_estimates

    @returns The CDE loss (up to a constant) for the CDE estimator on
    the holdout data and the SE error
    """

    n_obs, n_grid = cde_estimates.shape

    term1 = np.trapz(cde_estimates ** 2, z_grid.flatten())

    nns = [np.argmin(np.abs(z_grid - z_test[ii])) for ii in range(n_obs)]
    term2 = cde_estimates[range(n_obs), nns]

    loss = np.mean(term1 - 2 * term2)
    se_error = np.std(term1 - 2 * term2, axis=0) / (n_obs**0.5)

    return loss, se_error



### TEST
#' z_grid = np.linspace(0, 1, 99)
#' dist_unif = stats.uniform()
#' z_test = dist_unif.rvs(101)
#   dist = stats.beta(2,2)
#   cdes = np.array([dist.pdf(z_grid) for _ in range(101)]).reshape(101, 99)
#' cde_loss(cde, z_grid, z_test)